src/pool/srv_pool.c

```diff
@@ -3600,6 +3723,314 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+	daos_prop_free(prop);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	rc1 = rdb_tx_commit(&tx);
+out_tx:
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	if (need_put_leader)
+		pool_svc_put_leader(svc);
+
+	return rc1;
+}
+
+static int
+ds_pool_upgrade_if_needed(uuid_t pool_uuid, struct rsvc_hint *po_hint,
+			  struct pool_svc *svc, crt_rpc_t *rpc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			upgrade_status;
+	uint32_t			global_version;
+	int				rc, rc1;
+	bool				upgraded = false;
+	bool				need_put_leader = false;
+
+	if (!svc) {
+		rc = pool_svc_lookup_leader(pool_uuid, &svc, po_hint);
+		if (rc != 0)
+			return rc;
+		need_put_leader = true;
+	}
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	d_iov_set(&value, &global_version, sizeof(global_version));
+	rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_global_version,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_tx, rc);
+	} else if (rc == -DER_NONEXIST) {
+		if (!need_put_leader)
+			D_GOTO(out_tx, rc = 0);
+		D_GOTO(out_upgrade, rc);
+	} else {
+		d_iov_set(&value, &upgrade_status, sizeof(upgrade_status));
+		rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+				   &value);
+		if (rc)
+			D_GOTO(out_tx, rc);
+
+		if (global_version > DAOS_META_GLOBAL_VERSION) {
+			D_ERROR("downgrading pool is unsupported\n");
+			D_GOTO(out_tx, rc = -DER_INVAL);
+		}
+		switch (upgrade_status) {
+		case DAOS_UPGRADE_STATUS_NOT_STARTED:
+		case DAOS_UPGRADE_STATUS_COMPLETED:
+			if (global_version < DAOS_META_GLOBAL_VERSION)
+				D_GOTO(out_upgrade, rc = 0);
+			else
+				D_GOTO(out_tx, rc = 0);
+			break;
+		case DAOS_UPGRADE_STATUS_FAILED:
+			if (global_version < DAOS_META_GLOBAL_VERSION) {
+				D_ERROR("upgrading failed upgraded pool again unsupported\n");
+				D_GOTO(out_tx, rc = -DER_INVAL);
+			}
+			/* try again as users requested. */
+			D_GOTO(out_upgrade, rc = 0);
+			break;
+		case DAOS_UPGRADE_STATUS_IN_PROGRESS:
+			if (global_version < DAOS_META_GLOBAL_VERSION) {
+				D_ERROR("need finish last time upgraing\n");
+				D_GOTO(out_tx, rc = -DER_INVAL);
+			} else if (need_put_leader) { /* not from resume */
+				D_GOTO(out_tx, rc = -DER_INPROGRESS);
+			} else {
+				D_GOTO(out_upgrade, rc = 0);
+			}
+			break;
+		default:
+			D_ERROR("unknown upgrade pool status: %u\n", upgrade_status);
+			D_GOTO(out_upgrade, rc = -DER_INVAL);
+			break;
+		}
+	}
+out_upgrade:
+	/**
+	 * todo: make sure no rebuild/reint/expand are in progress
+	 */
+	rc = pool_upgrade_props(&tx, svc, pool_uuid, rpc);
+	if (rc == 0)
+		upgraded = true;
+out_tx:
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	if (need_put_leader) {
+		ds_rsvc_set_hint(&svc->ps_rsvc, po_hint);
+		pool_svc_put_leader(svc);
+	}
+
+	if (upgraded) {
+		rc1 = ds_cont_upgrade(pool_uuid, svc->ps_cont_svc);
+		rc = ds_pool_mark_upgrade_completed(pool_uuid, svc, rc1);
```

Comment:
If both need_put_leader and upgraded are true, it looks like we are using svc without a reference.